### PR TITLE
feat: in query

### PR DESCRIPTION
**Pull request by QuantumExplorer**

### CODE REVIEW

**Comment by QuantumExplorer** [comment]
The formatting will be on next PR right after this one.

File: Cargo.toml [hunk 1]
```diff
@@ -14,21 +14,16 @@ rand = "0.8.4"
...
 git = "https://github.com/dashevo/grovedb"
-branch = "master"
-
-# [dependencies.minicbor]
-# version = "0.12.0"
-# features = ["std"]
-
-# [dependencies.minicbor-derive]
-# version = "0.8.0"
+branch = "fix/iteration"
```
Comment (PastaPastaPasta) [inline Cargo.toml:26]: was this intended?

File: src/contract/mod.rs [hunk 2]
```diff
@@ -50,32 +50,48 @@ impl Index {
...
+                    // the return value for sort by age would be 1
+                    d -= 1;
+                }
+            }
+        }
+
+        // the in field can only be on the last or before last property
+        if in_field_name.is_some() {
+
+            if last_property.unwrap().name.as_str() != in_field_name.unwrap() {
```
Comment (PastaPastaPasta) [inline src/contract/mod.rs:79]: these if scopes should be combined

File: src/contract/mod.rs [hunk 3]
```diff
@@ -50,32 +50,48 @@ impl Index {
...
-                        // we can remove the -1 here
-                        // this is a case for example if we have an index on person's name and age
-                        // where we say name == 'Sam' sort by age
-                        // there is no field operator on age
-                        // The return value for name == 'Sam' sort by age would be 0
-                        // The return value for name == 'Sam and age > 5 sort by age would be 0
-                        // the return value for sort by age would be 1
-                        d -= 1;
-                    }
+                if !index_names.iter().any(|&a| a == sort_on.unwrap()) {
```
Comment (PastaPastaPasta) [inline src/contract/mod.rs:63]: combine to else if

File: src/contract/mod.rs [hunk 4]
```diff
@@ -317,34 +338,73 @@ impl DocumentType {
...
-            Error::CorruptedData(String::from("unable to get document id")),
-        )?;
-        document.remove("$id");
+        let mut document: HashMap<String, CborValue> =
+            ciborium::de::from_reader(read_document_cbor)
+                .map_err(|_| Error::CorruptedData(String::from("unable to decode contract")))?;
+
+        let owner_id = match owner_id {
+            None => {
+                let owner_id: Vec<u8> = bytes_for_system_value_from_hash_map(&document, "$ownerId")
```
Comment (PastaPastaPasta) [inline src/contract/mod.rs:360]: nit I'd prefer not shadowing here

File: src/contract/mod.rs [hunk 5]
```diff
@@ -356,13 +416,16 @@ impl Document {
...
         // first we need to deserialize the document and contract indices
         // we would need dedicated deserialization functions based on the document type
-        let mut document: HashMap<String, CborValue> = ciborium::de::from_reader(read_document_cbor)
-            .map_err(|_| Error::CorruptedData(String::from("unable to decode contract")))?;
+        let mut document: HashMap<String, CborValue> =
+            ciborium::de::from_reader(read_document_cbor)
+                .map_err(|_| Error::CorruptedData(String::from("unable to decode contract")))?;
 
         // dev-note: properties is everything other than the id
         let document = Document {
```
Comment (PastaPastaPasta) [inline src/contract/mod.rs:431]: double declaration of document is a mistake no?

File: src/contract/mod.rs [hunk 6]
```diff
@@ -382,18 +445,19 @@ impl Document {
...
-                let document_type =
-                    contract
-                        .document_types
-                        .get(document_type_name)
-                        .ok_or(Error::CorruptedData(String::from(
-                            "document type should exist for name",
-                        )))?;
-                Ok(Some(document_type.serialize_value_for_key(key, value)?))
+        if key == "$ownerId" && owner_id.is_some() {
+            Ok(Some(Vec::from(owner_id.unwrap())))
```
Comment (PastaPastaPasta) [inline src/contract/mod.rs:451]: this is a redundant unwrap. can you do smth like
```suggestion
        if key == "$ownerId" && let Some(owner_id) = unwrapped_owner_id {
            Ok(Some(Vec::from(unwrapped_owner_id)))
```

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

File: tests/query_tests.rs [hunk 7]
```diff
@@ -1,93 +1,319 @@
...
+        ]
+    });
+    let where_cbor = common::value_to_cbor(query_value, None);
+    let person_document_type = contract
+        .document_types
+        .get("person")
+        .expect("contract should have a person document type");
+    let query = DriveQuery::from_cbor(where_cbor.as_slice(), &contract, &person_document_type)
+        .expect("query should be built");
+    let (results, skipped) = query
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:284]: drop unused variable

```suggestion
    let (results, _) = query
```

File: tests/query_tests.rs [hunk 8]
```diff
@@ -1,93 +1,319 @@
...
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:246]: drop unused variable
```suggestion
    let (results, _) = query
```

File: tests/query_tests.rs [hunk 9]
```diff
@@ -1,93 +1,319 @@
...
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:172]: drop unused var
```suggestion
    let (results, _) = query
```

File: tests/query_tests.rs [hunk 10]
```diff
@@ -1,93 +1,319 @@
...
     }
     drive.grove.commit_transaction(db_transaction);
     (drive, contract)
 }
 
 #[test]
 fn test_query_many() {
-    let (mut drive, contract) = setup();
+    let (mut drive, contract) = setup(10, 73509);
+    let all_names = vec!["Adey".to_string(), "Briney".to_string(), "Cammi".to_string(), "Celinda".to_string(), "Dalia".to_string(), "Gilligan".to_string(), "Kevina".to_string(), "Meta".to_string(), "Noellyn".to_string(), "Prissie".to_string()];
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:94]: this is unused
```suggestion
    // let all_names = vec!["Adey".to_string(), "Briney".to_string(), "Cammi".to_string(), "Celinda".to_string(), "Dalia".to_string(), "Gilligan".to_string(), "Kevina".to_string(), "Meta".to_string(), "Noellyn".to_string(), "Prissie".to_string()];
```

File: tests/query_tests.rs [hunk 11]
```diff
@@ -1,93 +1,319 @@
-use grovedb::Query;
-use rand::Rng;
-use tempdir::TempDir;
+use std::collections::HashMap;
+use grovedb::{Error, Query};
+use rand::seq::SliceRandom;
+use rand::{Rng, SeedableRng};
+use rand_distr::{Exp, Distribution};
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:5]: Unused include
```suggestion
```

File: src/contract/mod.rs [hunk 12]
```diff
@@ -5,7 +5,7 @@ use ciborium::value::{Value as CborValue, Value};
 use grovedb::Error;
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
-use std::ops::Not;
+use array_tool::vec::Intersect;
```
Comment (PastaPastaPasta) [inline src/contract/mod.rs:8]: this is unused

```suggestion
```

File: tests/common/mod.rs [hunk 13]
```diff
@@ -1,11 +1,11 @@
+use byteorder::{BigEndian, WriteBytesExt};
+use rs_drive::contract::Contract;
+use rs_drive::drive::Drive;
 use std::fs::File;
 use std::io;
 use std::io::{BufRead, BufReader, Write};
```
Comment (PastaPastaPasta) [inline tests/common/mod.rs:6]: nit this is unused
```suggestion
use std::io::{BufRead, BufReader};
```

File: tests/query_tests.rs [hunk 14]
```diff
@@ -1,93 +1,319 @@
...
 use rs_drive::contract::{Contract, Document, DocumentType};
 use rs_drive::drive::Drive;
 use rs_drive::query::DriveQuery;
 use serde::{Deserialize, Serialize};
-use rand::seq::SliceRandom;
 use serde_json::json;
+use tempdir::TempDir;
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:11]: this is unused
```suggestion
```

File: tests/query_tests.rs [hunk 15]
```diff
@@ -1,93 +1,319 @@
-use grovedb::Query;
-use rand::Rng;
-use tempdir::TempDir;
+use std::collections::HashMap;
+use grovedb::{Error, Query};
+use rand::seq::SliceRandom;
+use rand::{Rng, SeedableRng};
+use rand_distr::{Exp, Distribution};
 use rs_drive::contract::{Contract, Document, DocumentType};
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:6]: nit: unused
```suggestion
use rs_drive::contract::{Contract, Document};
```

File: tests/query_tests.rs [hunk 16]
```diff
@@ -1,93 +1,319 @@
-use grovedb::Query;
-use rand::Rng;
-use tempdir::TempDir;
+use std::collections::HashMap;
+use grovedb::{Error, Query};
```
Comment (PastaPastaPasta) [inline tests/query_tests.rs:2]: unused 
```suggestion
```

File: src/query/mod.rs [hunk 17]
```diff
@@ -473,29 +546,45 @@ impl<'a> DriveQuery<'a> {
...
+        let mut fields = equal_fields;
+        let mut sort_on = None;
+        if range_field.is_some() {
+            fields.push(range_field.unwrap());
+            sort_on = Some(range_field.unwrap());
+        }
+        if in_field.is_some() {
+            fields.push(in_field.unwrap());
+            //if there is an in_field, it always takes precedence
+            sort_on = Some(in_field.unwrap());
```
Comment (PastaPastaPasta) [inline src/query/mod.rs:582]: `sort_on` is never read from?

File: src/drive/mod.rs [hunk 18]
```diff
@@ -1,7 +1,7 @@
 pub mod defaults;
 
 use crate::contract::{Contract, Document};
-use ciborium::value::{Value as CborValue};
+use ciborium::value::Value as CborValue;
```
Comment (PastaPastaPasta) [inline src/drive/mod.rs:4]: this is unused
```suggestion
```

File: src/contract/mod.rs (hunk 6 above)
Comment (PastaPastaPasta) [reply src/contract/mod.rs:451]: nevermind, this isn't easy to do here :)

**Pull request merged**